sql/rpl_master_info_file.hh

```diff
@@ -0,0 +1,613 @@
+    virtual bool load_from(IO_CACHE *file) override
+    { return IntIOCache::from_chars<I>(file, this); }
+    virtual void save_to(IO_CACHE *file) override
+    { return IntIOCache::to_chars(file, operator I()); }
+  };
+
+  /** SSL Path Field:
+    @ref FN_REFLEN-sized C-string with a `mariadbd` option for the `DEFAULT`.
+    Empty string is "\0\0" and `DEFAULT`ed string is "\0\1".
+  */
+  template<const char *const &mariadbd_option>
+  struct OptionalPathField: StringField<>
+  {
+    operator const char *() override
+    {
+      if (is_default())
+        return mariadbd_option;
+      return StringField<>::operator const char *();
+    }
+    auto &operator=(const char *other)
+    {
+      buf[1]= false; // not default
+      StringField<>::operator=(other);
+      return *this;
+    }
+    bool is_default() override { return !buf[0] && buf[1]; }
+    bool set_default() override
+    {
+      buf[0]= false;
+      buf[1]= true;
+      return false;
+    }
+    bool load_from(IO_CACHE *file) override
+    {
+      buf[1]= false; // not default
+      return StringField<>::load_from(file);
+    }
+  };
+
+  /** Boolean Field with `DEFAULT`:
+    This uses a trilean enum,
+    which is more efficient than `std::optional<bool>`.
+    load_from() and save_to() are also engineered
+    to make use of the range of only two cases.
+  */
+  template<bool &mariadbd_option> struct OptionalBoolField: Persistent
+  {
+    enum { NO, YES, DEFAULT= -1 } value;
+    operator bool() { return is_default() ? mariadbd_option : (value != NO); }
+    bool is_default() override { return value <= DEFAULT; }
+    bool set_default() override
+    {
+      value= DEFAULT;
+      return false;
+    }
+    auto &operator=(bool value)
+    {
+      this->value= value ? YES : NO;
+      return *this;
+    }
+    /// @return `true` if the line is `0` or `1`, `false` otherwise or on error
+    bool load_from(IO_CACHE *file) override
+    {
+      /** Only three chars are required:
+        * One digit
+          (When base prefixes are not recognized in integer parsing,
+          anything with a leading `0` stops parsing
+          after converting the `0` to zero anyway.)
+        * the terminating `\n\0` as in IntegerLike::from_chars(IO_CACHE *, I &)
+      */
+      char buf[3];
+      if (my_b_gets(file, buf, 3) && buf[1] == '\n')
+        switch (buf[0]) {
+        case '0':
+          value= NO;
+          return false;
+        case '1':
+          value= YES;
+          return false;
+        }
+      return true;
+    }
+    void save_to(IO_CACHE *file) override
+    { my_b_write_byte(file, operator bool() ? '1' : '0'); }
+  };
+
+
+  /** @ref uint32_t Array field
+    @deprecated
+      Only one of `DO_DOMAIN_IDS` and `IGNORE_DOMAIN_IDS` can be active
+      at a time, so giving them separate arrays, let alone field instances,
+      is wasteful. Until we refactor this pair, this will only reference
```

Comment:
I first brought this up [on Zulip](https://mariadb.zulipchat.com/#narrow/channel/118759-general/topic/Replication/near/548006162).